src/ui/screen_process.c

```diff
@@ -46,52 +60,44 @@ static component_t* _get_waiting_screen(void)
     return waiting_screen;
 }
 
-static void _screen_process(bool (*is_done)(void), void (*on_timeout)(void), const uint32_t timeout)
+/**
+ * Renders the provided component on the display.
+ *
+ * @param[in] component Screen to draw.
+ */
+static void _screen_draw(component_t* component)
 {
-    component_t* waiting_screen = _get_waiting_screen();
-    uint32_t timeout_cnt = 0;
-
-    bool screen_new = false;
-    component_t* component = NULL;
-    while (is_done == NULL || !is_done()) {
-        component_t* next_component = ui_screen_stack_top();
-        if (next_component == NULL) {
-            next_component = waiting_screen;
-        }
-        screen_new = false;
-        if (next_component != component) {
-            screen_new = true;
-            component = next_component;
-        }
-        gestures_detect(screen_new, component->emit_without_release);
-        if (screen_frame_cnt == SCREEN_FRAME_RATE) {
-            if (is_done != NULL && on_timeout != NULL && timeout_cnt > timeout) {
-                on_timeout();
-            }
-            screen_frame_cnt = 0;
-            timeout_cnt += 1;
-            ui_screen_render_component(component);
-        }
-        screen_frame_cnt++;
-        ui_screen_stack_cleanup();
-        if (is_done == NULL) {
-            usb_processing_process(usb_processing_hww());
-#if defined(APP_U2F)
-            usb_processing_process(usb_processing_u2f());
-#endif
-        }
+    if (screen_frame_cnt == SCREEN_FRAME_RATE) {
+        screen_frame_cnt = 0;
+        ui_screen_render_component(component);
     }
+    screen_frame_cnt++;
 }
 
-void ui_screen_process(bool (*is_done)(void))
+/*
+ * Select which activity we should draw next
+ * (or fallback to the idle screen).
+ */
+static component_t* _get_ui_top_component(void)
 {
-    _screen_process(is_done, NULL, 0);
+    component_t* result = ui_screen_stack_top();
+    if (!result) {
+        return _get_waiting_screen();
+    }
+    return result;
 }
 
-void ui_screen_process_with_timeout(
-    bool (*is_done)(void),
-    void (*on_timeout)(void),
-    uint32_t timeout)
+void ui_spin(void)
```

Comment:
this function should be called `screen_process_spin` (usually same as filename).

can rename the whole file and the functions in here in a separate PR, but would keep `screen_process_*` now to keep consistent.

furthermore, `spin` to me means it is looping, but it is not looping. Please find a better name (`screen_process()` ?)

Reply:
Fixed. `screen_process` wins.